src/caffeine/teams_s.f90

```diff
@@ -1,20 +1,70 @@
 ! Copyright (c), The Regents of the University of California
 ! Terms of use are as specified in LICENSE.txt
 submodule(prif:prif_private_s) teams_s
-  implicit none
+  use iso_c_binding, only: c_null_funptr, c_f_pointer, c_loc
 
+  implicit none
 contains
 
   module procedure prif_change_team
-    call unimplemented("prif_change_team")
+    if (caf_this_image(team%info%gex_team) == 1) then ! need to setup the heap for the team
+      team%info%heap_start = current_team%info%child_heap_info%offset + current_team%info%heap_start
+      team%info%heap_size = current_team%info%child_heap_info%size
+      call caf_establish_mspace( &
+          team%info%heap_mspace, &
+          as_c_ptr(team%info%heap_start), &
+          current_team%info%child_heap_info%size)
+    end if
+    current_team = team
+    if (caf_have_child_teams()) then ! need to establish heap for child teams
+      call caf_establish_child_heap
+    end if
   end procedure
 
   module procedure prif_end_team
-    call unimplemented("prif_end_team")
+    type(prif_coarray_handle), allocatable :: teams_coarrays(:)
+    integer :: num_coarrays_in_team, i
+    type(handle_data), pointer :: tmp_data
+
+    ! deallocate the teams coarrays
```

Comment:
The two-pass algorithm with a heap temporary below could instead have been written (more simply) as a one-pass that deallocates each coarray individually. 

I assume you wrote it this way in order to batch together multiple deallocations into a single `prif_deallocate_coarray()` call?

This seems like the right choice, but it's non-obvious at first glance. Request a comment to explain this, eg:

```suggestion
    ! deallocate the teams coarrays
    ! Currently we work to batch together all the deallocations into a single call
    ! to prif_deallocate_coarray(), in the hope it can amortize some costs
```

Reply:
You're right, I like your suggested comment so I added it.